### DISCUSSION TITLE
DS18x20 ids over reboot

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19021] [General]
Are generated ID's for DS18x20  in any way correlated to the 1wire address, so that after reboot the same address gets the same id?

**Comment by barbudor** [comment]
Not when using the default ID
To get persistent identification you need to compile your own binary with one of those 2 options
- DS18x20_USE_ID_AS_NAME to use the mower part of the 1-wire address as an ID
- DS18x20_USE_ID_ALIAS to have the ability to make your own mapping between the 1-wire address and a numeric ID

More details in the docs here https://tasmota.github.io/docs/DS18x20/#compile-options

**Reply by sfromis** [reply]
Or you can use the persistent `Id` already provided by the hardware, instead of using the `DS18B20_xx` number.

**Reply by msetina** [reply]
@sfromis I was wondering how can I get this in normal firmwares and provided driver. In documentation I could not find this. 
If it was accesible in scripting (Berry) it would be good enought.

**Reply by sfromis** [reply]
Yes, the `Id` field is available in the published data sensor like this:
`{"Time":"2023-07-03T20:08:31.810+02:00","DS18B20-1":{"Id":"0304979453B2","Temperature":22.00},"TempUnit":"C"}`

Same exact payload is available to Berry using `tasmota.read_sensors()`, and then you can parse it like this:
```
import json
json.load(tasmota.read_sensors())
{'DS18B20-1': {'Id': '0304979453B2', 'Temperature': 22.00}, 'Time': '2023-07-03T20:11:57.837+02:00', 'TempUnit': 'C'}
```
You can access fields of this nested map using usual Berry code for traversing the fields, like looping trough the objects to look for the `id` you're interested in.

Where you do not get the `Id` is on the Tasmota web UI front page, where you only see the not-stable number, unless you do a custom build using one of the features mentioned by barbudor.

**Reply by msetina** [reply]
This is OK, if Id is the hardware address, my Driver can provide WebUI with data that is good enough to identify a sensor in real world.

**Reply by sfromis** [reply]
Yes, `Id` is what the hardware says. I'm not 100% convinced that all producers of [clones of "varying quality"](https://github.com/cpetrich/counterfeit_DS18B20) are strict about guaranteeing unique values, but Tasmota is not inventing anything.....

**Comment by msetina** [comment]
From documentation it is not clear if Id in JSON is the hardware address and if it is available in tasmota.read_sensors()

**Reply by barbudor** [reply]
Yes to both questions

**Comment by msetina** [comment]
Is there specail encoding of id number, because I can not connect info I get in tasmota comparing to python:
```
tasmota
0822547C02B2
0722614B5AA5
072261464817

python 1wire:
28a55a4b61220728
2817484661220723

**Reply by sfromis** [reply]
Looks like just a segment, and different byte ordering.
`072261464817` is contained within `28 174846612207 23` 

The formatting to the hex `Id` skips sensor type and crc, that's the start and end bytes.

**Reply by msetina** [reply]
Kudos. Thank you. Will test this hypothesis......

**Reply by msetina** [reply]
This checks out, but ... As engineers are humans, there are now two different notations of the id. One on  Linux and micropython and one on tasmota and arduino. OK, arduino adds device type ident 0x28, tasmota has it in a name (18b20-1). The difference is byte order and CRC.

**Reply by sfromis** [reply]
As Tasmota is sitting on top of Arduino Core, a large part of the device drivers are based on, or includes, device drivers created for Arduino. The current code is explicitly reversing byte order when hexifying, likely to match what already was done with Arduino.
